File: JAVA/JumpList.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

class Node {
    int key;
    Node next;
    Node prev;
    Node down;

    Node(int key) {
        this.key = key;
        this.next = null;
        this.prev = null;
        this.down = null;
    }
}

class LevelList {
    Node head;
    Node tail;

    LevelList() {
        head = new Node(-1);       // -∞ sentinel
        tail = new Node(10000);    // +∞ sentinel
        head.next = tail;
        tail.prev = head;
    }
}

public class JumpList {
    List<LevelList> levels;
    Random rand;

    public JumpList() {
        levels = new ArrayList<>();
        levels.add(new LevelList());  // Level 0
        rand = new Random();
    }

    public void addLevel() {
        LevelList newLevel = new LevelList();
        newLevel.head.down = levels.get(levels.size() - 1).head;
        newLevel.tail.down = levels.get(levels.size() - 1).tail;
        levels.add(newLevel);
    }

    public void insert(int key) {
        int level = 0;
        while (rand.nextInt(2) == 0) level++;
        while (levels.size() <= level) addLevel();

        Node newNode = new Node(key);
        Node current = levels.get(level).head;

        // Find insert position on top level
        while (current.next.key < key) {
            current = current.next;
        }

        // Insert node
        newNode.next = current.next;
        newNode.prev = current;
        current.next.prev = newNode;
        current.next = newNode;

        // Insert downward recursively
        insertDown(newNode, newNode.prev.down, key);
    }

    private void insertDown(Node upperNode, Node downNode, int key) {
        if (downNode == null) return;

        Node newNode = new Node(key);
        upperNode.down = newNode;

        Node current = downNode;
        while (current.next.key < key) {
            current = current.next;
        }

        newNode.next = current.next;
        newNode.prev = current;
        current.next.prev = newNode;
        current.next = newNode;

        insertDown(newNode, newNode.prev.down, key);
    }

    public Node search(int key) {
        Node current = levels.get(levels.size() - 1).head;
        while (current != null) {
            while (current.next != null && current.next.key <= key) {
                current = current.next;
            }
            if (current.key == key) return current;
            current = current.down;
        }
        return null;
    }

    public boolean delete(int key) {
        Node current = levels.get(levels.size() - 1).head;
        boolean found = false;

        while (current != null) {
            while (current.next != null && current.next.key < key) {
                current = current.next;
            }

            if (current.next != null && current.next.key == key) {
                Node target = current.next;
                current.next = target.next;
                target.next.prev = current;
                found = true;
                current = current.down;
            } else {
                current = current.down;
            }
        }

        return found;
    }

    // Optional: print skip list from top to bottom
    public void print() {
        for (int i = levels.size() - 1; i >= 0; i--) {
            Node current = levels.get(i).head;
            System.out.print("Level " + i + ": ");
            while (current != null) {
                System.out.print(current.key + " ");
                current = current.next;
                if (current != null && current.key == 10000) break;
            }
            System.out.println();
        }
    }
}

File: C++/JumpList.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;

struct Node {
    int key;
    Node* next;
    Node* prev;
    Node* down;

    Node(int k) : key(k), next(nullptr), prev(nullptr), down(nullptr) {}
};

struct LevelList {
    Node* head;
    Node* tail;

    LevelList() {
        head = new Node(-1);     // -∞
        tail = new Node(10000);  // +∞
        head->next = tail;
        tail->prev = head;
    }
};

class SkipList {
public:
    vector<LevelList*> levels;

    SkipList() {
        levels.push_back(new LevelList());  // Level 0
        srand((unsigned int)time(nullptr)); // 初始化随机种子
    }

    ~SkipList() {
        for (LevelList* level : levels) {
            Node* current = level->head;
            while (current) {
                Node* next = current->next;
                delete current;
                current = next;
            }
            delete level;
        }
    }

    void addLevel() {
        LevelList* newLevel = new LevelList();
        newLevel->head->down = levels.back()->head;
        newLevel->tail->down = levels.back()->tail;
        levels.push_back(newLevel);
    }

    void insert(int key) {
        int level = 0;
        while (rand() % 2 == 0) level++;
        while ((int)levels.size() <= level) addLevel();

        Node* newNode = new Node(key);
        Node* current = levels[level]->head;

        // 寻找插入位置（最上层）
        while (current->next->key < key) {
            current = current->next;
        }

        // 插入当前层
        newNode->next = current->next;
        newNode->prev = current;
        current->next->prev = newNode;
        current->next = newNode;

        // 向下插入
        insertDown(newNode, newNode->prev->down, key);
    }

    void insertDown(Node* upperNode, Node* downNode, int key) {
        if (!downNode) return;

        Node* newNode = new Node(key);
        upperNode->down = newNode;

        Node* current = downNode;
        while (current->next->key < key) {
            current = current->next;
        }

        newNode->next = current->next;
        newNode->prev = current;
        current->next->prev = newNode;
        current->next = newNode;

        insertDown(newNode, newNode->prev->down, key);
    }

    Node* search(int key) {
        Node* current = levels.back()->head;
        while (current) {
            while (current->next && current->next->key <= key) {
                current = current->next;
            }
            if (current->key == key) return current;
            current = current->down;
        }
        return nullptr;
    }

    bool deleteKey(int key) {
        Node* current = levels.back()->head;
        bool found = false;

        while (current) {
            while (current->next && current->next->key < key) {
                current = current->next;
            }

            if (current->next && current->next->key == key) {
                Node* target = current->next;
                current->next = target->next;
                target->next->prev = current;
                delete target;
                found = true;
                current = current->down;
            } else {
                current = current->down;
            }
        }

        return found;
    }

    void print() {
        for (int i = (int)levels.size() - 1; i >= 0; --i) {
            Node* current = levels[i]->head;
            cout << "Level " << i << ": ";
            while (current) {
                cout << current->key << " ";
                current = current->next;
                if (current && current->key == 10000) break;
            }
            cout << endl;
        }
    }
};

// === 测试主函数 ===
int main() {
    SkipList sl;
    for (int i = 1; i <= 10; ++i) {
        sl.insert(i * 10);
    }

    sl.print();

    int key = 50;
    cout << "Searching for " << key << ": " << (sl.search(key) != nullptr ? "Found" : "Not found") << endl;
    cout << "Deleting " << key << ": " << (sl.deleteKey(key) ? "Deleted" : "Not found") << endl;
    sl.print();

    return 0;
}
